evolver/hardware/standard/temperature.py

```diff
@@ -0,0 +1,64 @@
+from copy import copy
+
+from pydantic import Field
+
+from evolver.hardware.interface import (
+    EffectorDriver,
+    SensorDriver,
+    VialBaseModel,
+)
+from evolver.hardware.standard.base import SerialDeviceConfigBase, SerialDeviceOutputBase
+from evolver.serial import SerialData
+
+
+class Temperature(SensorDriver, EffectorDriver):
+    """Temperature sensor and heater package.
+
+    Goes over the evolver serial protocol and is capable of both reading the current
+    temperature value as well as setting a target for heating. A particular raw value
+    stands in for off (TEMP_OFF property of this class).
+    """
+
+    TEMP_OFF = b"4095"
+
+    class Config(SerialDeviceConfigBase):
+        pass
+
+    class Output(SerialDeviceOutputBase):
+        raw: int
+        temperature: float = Field(None, description="Sensor temperature in degrees celcius")
+
+    class Input(VialBaseModel):
+        temperature: float = Field(None, description="Target temperature in degrees celcius")
+
+    @property
+    def serial(self):
+        return self.serial_conn or self.evolver.serial
+
+    def _do_serial(self, from_proposal=False):
+        data = [self.TEMP_OFF] * self.slots
+        # since a read is also a send, we load all committed values as a base and
+        # in the case of proposals overwrite with new data.
+        inputs = copy(self.committed)
+        if from_proposal:
+            inputs.update(self.proposal)
+        for vial, input in inputs.items():
+            if vial in self.vials:
+                # calibration from real to raw should go here
+                raw = int(input.temperature)
+                data[vial] = str(raw).encode()
+        with self.serial as comm:
+            response = comm.communicate(SerialData(addr=self.addr, data=data))
+        self.committed = inputs
```

Comment:
The committed stores the real values (here "Input.temperature" which is in degrees C). At the moment since we don't have the calibration it just fakes it, but it ought to do the calibration from inputs (or previous committed) prior to sending command. Because the inputs could be a mix of new and previously committed, it should be simpler just to compute the conversion each time.

committed is a useful structure since there seems to be no way to ask from the underlying device what has been set, so this could be queries by a controller to get that info.